### ISSUE TITLE
Race condition leading to Validation Failed: {"resource":"Release","code":"already_exists","field":"tag_name"}

### ISSUE THREAD

**Issue by rhazarian** [ncipollo/release-action#448]
The problem occurs when you have a matrix job which tries to upload its artifacts to a GH release (with `allowUpdates: true`, `replacesArtifacts: false`). The expected behavior is that the first finished sub-task will create the release, and then the subsequent sub-tasks will upload additional artifacts to the release.

However, what sometimes happen is:
1. Sub-tasks A and B finish almost simultaneously
2. Sub-task A checks if the release already exists - gets False
3. Sub-task B checks if the release already exists - gets False
4. Sub-task A tries to create the release - the release is successfully created
5. Sub-task B tries to create the release - bah, `Error: Error 422: Validation Failed: {"resource":"Release","code":"already_exists","field":"tag_name"} ncipollo/release-action@v1`, the whole job fails

I guess this error should be re-tried at least once to eliminate the possible race condition.

**Comment by ncipollo** [comment]
As you allude to, you are creating a race condition in your workflow. Resolving that race condition is out of scope for this action. Your best bet is to have the release action steps happen in sequence.

**Comment by rhazarian** [comment]
> As you allude to, you are creating a race condition in your workflow. Resolving that race condition is out of scope for this action. Your best bet is to have the release action steps happen in sequence.

Why is it out of its scope though? The action's contract is 'create the release if it doesn't exist yet, update it if it exists'. It should be an atomic operation, the fact that it is done in 2 separate API calls is an implementation detail which is not even documented.

**Comment by rhazarian** [comment]
> Your best bet is to have the release action steps happen in sequence.

If you are building the same thing for multiple targets, you usually do it in parallel, no one does it in sequence. (If your build takes 10 minutes and you need to build for 5 different targets, you'd want your build to complete in 10 minutes, not in 50).

To avoid the race condition without modifications to this action, you'd need to upload all of the artifacts from all of the parallel builds to a temporary storage, then pull them in a separate step and then upload them to a release. Which is slower and fairly cumbersome.

However the whole problem could be easily & efficiently solved with a single retry for the `already_exists` error code. Could you elaborate why are you strictly against that?

**Comment by ncipollo** [comment]
Even if the release creation or update could be atomic you would still have an issue in your workflow. Both task A and B can take an arbitrary amount of time when run in parallel. It is impossible for you to guarantee the order you want without some kind of synchronization mechanism. I don't think it makes sense to add retry or some kind of syncing mechanism in this action to accommodate your workflow, however there is probably a way within the workflow itself to achieve what you are after (for example maybe something like this - https://github.com/marketplace/actions/wait-for-jobs)

**Comment by rhazarian** [comment]
> Even if the release creation or update could be atomic you would still have an issue in your workflow. Both task A and B can take an arbitrary amount of time when run in parallel. It is impossible for you to guarantee the order you want without some kind of synchronization mechanism. 

In this case, the order of the matrix job sub-tasks doesn't matter.

Here is what it supposed to happen:

- 5 jobs run in parallel. Each of them produces an artifact, which is uploaded to a draft release. If the release doesn't exist yet, it should be created.
- There is another job that waits for those 5 jobs. It is supposed that when all the 5 parallel jobs are finished successfully, a draft release should exist with all of the 5 artifacts attached (the order in which they were attached does not matter). Then, this final job publishes the draft release.

I would say that this is a fairly common case. Probably it's just not very usual that a couple of parallel jobs finish the same moment.

Now I release that the problem may be worked around by created an empty draft release prematurely (before running the 5 parallel jobs). However I still think that it would be nice if this action would handle it out-of-the-box.

**Comment by ncipollo** [comment]
> However the whole problem could be easily & efficiently solved with a single retry for the already_exists error code. Could you elaborate why are you strictly against that?

This action is used in many different kinds of workflow - it needs to have simple behavior which works generally in most cases. I don't think it makes sense for the action to assume that error means you are hitting a race condition vs it failing because you tried to manipulate the release in an unsupported way. 
 
Further I don't think this is the last issue you are going to hit with your current workflow. It is generally a bad idea to attempt to mutate a shared resource from multiple tasks in parallel. You are creating a non-deterministic series of events and it's going to be difficult to ensure your system works correctly in all combinations of timing.

**Comment by dreed-sd** [comment]
For with it's worth I ran into this exact problem and found this discussion searching for a solution. We build for multiple platforms and combine them into a single release. They build in parallel, and it works 95% of the time. Every now and then, this happens, and we have to restart the workflow. It's a shame it doesn't work, I now have to do something more complicated. I think you should at least update the documentation that the `allowUpdates` flag will not allow for this type of workflow, at least not reliably.
